sklearn/feature_selection/tests/test_from_model.py

```diff
@@ -40,6 +41,130 @@ def test_input_estimator_unchanged():
+
+
+def check_valid_max_features(est, X, y):
+    max_features = X.shape[1]
+    for valid_max_n_feature in [0, max_features, 'all', 5]:
+        transformer = SelectFromModel(estimator=est,
+                                      max_features=valid_max_n_feature,
+                                      threshold=-np.inf)
+        X_new = transformer.fit_transform(X, y)
+        if valid_max_n_feature == 'all':
+            valid_max_n_feature = max_features
+        assert_equal(X_new.shape[1], valid_max_n_feature)
+
+
+class FixedImportanceEstimator(BaseEstimator):
+    def __init__(self, importances):
+        self.importances = importances
+
+    def fit(self, X, y=None):
+        self.feature_importances_ = np.array(self.importances)
+
+
+def check_max_features(est, X, y):
+    X = X.copy()
+    max_features = X.shape[1]
+
+    check_valid_max_features(est, X, y)
+    check_invalid_max_features(est, X, y)
+
+    transformer1 = SelectFromModel(estimator=est, max_features='all',
+                                   threshold=-np.inf)
+    transformer2 = SelectFromModel(estimator=est,
+                                   max_features=max_features,
+                                   threshold=-np.inf)
+    X_new1 = transformer1.fit_transform(X, y)
+    X_new2 = transformer2.fit_transform(X, y)
+    assert_array_equal(X_new1, X_new2)
+
+    # Test max_features against actual model.
+    transformer1 = SelectFromModel(estimator=Lasso(alpha=0.025,
+                                   random_state=42))
+    X_new1 = transformer1.fit_transform(X, y)
+    scores1 = np.abs(transformer1.estimator_.coef_)
+    candidate_indices1 = np.argsort(-scores1, kind='mergesort')
+
+    for n_features in range(1, X_new1.shape[1] + 1):
+        transformer2 = SelectFromModel(estimator=Lasso(alpha=0.025,
+                                       random_state=42),
+                                       max_features=n_features,
+                                       threshold=-np.inf)
+        X_new2 = transformer2.fit_transform(X, y)
+        scores2 = np.abs(transformer2.estimator_.coef_)
+        candidate_indices2 = np.argsort(-scores2, kind='mergesort')
+        assert_array_equal(X[:, candidate_indices1[:n_features]],
+                           X[:, candidate_indices2[:n_features]])
+    assert_array_equal(transformer1.estimator_.coef_,
+                       transformer2.estimator_.coef_)
+
+    # Test if max_features can break tie among feature importance
+
+    feature_importances = np.array([4, 4, 4, 4, 3, 3, 3, 2, 2, 1])
+    for n_features in range(1, max_features + 1):
+        transformer = SelectFromModel(
+            FixedImportanceEstimator(feature_importances),
+            max_features=n_features,
+            threshold=-np.inf)
+        X_new = transformer.fit_transform(X, y)
+        selected_feature_indices = np.where(transformer._get_support_mask())[0]
+        assert_array_equal(selected_feature_indices, np.arange(n_features))
```

Comment:
I'm okay with this approach, but wonder if we'd be better off taking `max_features` literally and returning none of the tying features at the cutoff (to avoid users being surprised by the tie-breaking; although we do break ties like this in `SelectKBest` and `SelectPercentile`, and perhaps we should remain consistent). WDYT? Perhaps it rarely matters.

Reply:
I personally think it is better to give users exactly the number of features they ask for. From my experience as a user, I don't care so much which of feature X or feature Y I get when both are tied in importance as much as I do that when I ask for Z features, I get Z features and not less. Consistency with `SelectKBest` and `SelectPercentile` would be other arguments in favor of keeping as is. However, this is just my 2 cents, and I'll defer to you on the final decision.